### ISSUE TITLE
Reframe `MSRunSample` and `MSRunSequence` to be independent of the path between `PeakGroup`->`Sample`

### ISSUE THREAD

**Issue by hepcat72** [Princeton-LSI-ResearchComputing/tracebase#947]
### Goal

Make uploading `mzXML` files independent of uploading `accucor`/`isocorr` files.

### Inspiration

I want to suggest a change that I think is important to make that will make TraceBase more agile to future changes and simplify code necessary for high priority features we want to implement (specifically, the loading of `mzXML` files unassociated with accucor data).  These changes aren't *necessary*, but I think will reduce bugs, complexity, and make TraceBase more amenable to features built on top of what we have.  There are a lot of facets to consider, but the change I propose below is fairly conceptually simple.  In this inspiration section of this change request ("feature request" issue), I want to try and succinctly describe the problems that this feature request is designed to address...  After I submit this issue, I will proceed directly to implementing what was discussed in recent meetings as high priority: *the ability to load mzXML files unassociated with loaded accucor data* (#905).  I feel like doing this first would IMO be better, but given the potential for unaccounted-for considerations, I think it's feasible to move on to 905 and discuss this in parallel.

#### Existing issues with the models

I think that our original design (v1) was intended to represent the conceptual reality, which I think was good.  An `MSRun` was an attempt to represent what was happening on the instrument, but without some unique way of defining that, all we could do was in essence, create a placeholder and try and capture enough info to force it to be unique (person, day, protocol, and sample).  That concept continued through to v3.  We just split it into 2 models and added more data to try and be more granular: instrument, polarity, and scan range.  Having an mzXML and/or raw file is unique-enough for `MSRunSample`, but it's optional, and even with it, when we're loading metadata, the files are too large to be able to use them to compile LCMS data that is unique enough to detect issues - and creating records *without* them creates a lot of overhead.  The metadata without the file content is not unique enough for us to be able to accommodate unique "placeholder records".  While the representation of the placeholder records (absent mzXML) could be worked around, once we want to add an mzXML and associate it with an existing record, we could potentially have a complex web of associations to clean up.

#### Example

Let me see if I can convey what I mean by example.  Let's say we create an `MSRunSample` record like the following:

```
sample1=Sample(...)
lcm = LCMethod(...)
sequence = MSRunSequence(
    researcher="George",
    date="1/1/2024",
    instrument="QE2",
    lc_method= lcm,
)
msrsample = MSRunSample(
    msrun_sequence=sequence,
    sample=sample1,
    polarity="positive",
    mz_min=1.0,
    mz_max=500.0,
    ms_raw_file=None,
    ms_data_file=None,
)
af1 = ArchiveFile(
    ..."accucor1.xlsx"...
)
af2 = ArchiveFile(
    ..."accucor2.xlsx"...
)
proline_pg1 = PeakGroup(
    ...
    msrun_sample=msrsample,
    peak_annotation_file=af1,
    ...
)
proline_pg2 = PeakGroup(
    ...
    msrun_sample=msrsample,
    peak_annotation_file=af2,
    ...
)
```
But George did 2 runs on that day with the same protocol with the same set of samples, polarity, and overlapping scan ranges.  When we create these records, it doesn't matter too much that each one represents multiple actual events and that the 2 proline peak groups used different mxXML files.  Each MSRun record (sequence and sample) is a placeholder and we don't need to differentiate the records.  For all we know, the same mzXML went into both accucor file creations.  And that's not a problem, until...

George tries to add a subset of `mzXML` files for one of the 2 runs at some later point.  What do we do?

- We can just add the mzXML to the existing record and update it or...
- We can create a new (second) record.

Incidentally, right now, the code has a separate issue surrounding the assumption that the headers in the accucor files are unique, but that can be addressed on its own.

#### The real problems that creates...

The problems either strategy above creates are issues in and of themselves, and if we continue to build on top of those issues, we will end up with complex code and a higher incidence of bugs.  Here are some issues I can envision with each of the work-arounds above:

- We can just add the mzXML to the existing record and update it or...
  - Adding an mzXML will associate `proline_pg1` or `proline_pg2` with the wrong `mzXML`
  - We have no discrete way of knowing which accucor file is truly associated with the `mzXML` linked from the `MSRunSample` record is the file names and headers are the same.
  - Requiring users to associate mzXMLs with the accucor file they went into is:
    - time-consuming, intimidating, and difficult for the user
    - also unnecessary - we don't care for 1, but also, we don't even need to MSRunSample record in order to accurately link a column of accucor data to a sample. We can't discretely represent an MSRun (or injection event for that matter) to an accucor file or a peak group, and when we do make the association, it's really just an educated guess.
- We can create a new (second) record.
  - This creates overhead of maintaining multiple incomplete and complete `MSRunSample` records along-side one another
  - It also means that once we get 2 mzXMLs for the same sample, we would have to split the `MSRunSequence` record as well
  - And if there are more than 2 mzXMLs with the same polarity and scan range, when we get to a second or third file, do we maintain a placeholder record without an mzXML or not?

There are simply numerous ambiguities and judgement calls that can create confusion, potential bugs, inaccuracies, etc.  I think that we're just asking for problems down the road by not addressing these issues now.

### Description

Based on the above, I think we should think about `MSRunSequence` and `MSRunSample` differently.

An `MSRunSequence` record isn't a single "sequence".  It's a representation of potentially multiple sequences.  To put it (inaccurately for brevity), it's a "sequence protocol person day".  To make my point, we could even have (though I'm not suggesting this) an attribute for "number of sequences", e.g. `MSRunSequence.num_sequences` that were run on that day by the same person using the same protocol.

By the same token, what we're grappling with now, is how to maintain 2 representations of an `MSRunSample` record (with and without mzXML), and even when we have mzXMLs, since they are too big to move around when compiling data for a study, there is a state of an `MSRunSample` record where when we don't have the hash yet and multiple sample headers with the same value, essentially the same thing: the potential for an `MSRunSample` record that potentially represents multiple records.

And that is complicated by the fact that PeakGroups link directly to `MSRunSample` records.  We need that link currently, to get from a `PeakGroup` to a `Sample`.

The change I'm proposing is 3-fold, but let me take this one step at a time.  Let's sever our dependence on an `MSRunSample` record as the necessary link between `PeakGroup` (associated with an accucor file) and `Sample` by making these changes:

- Change: `PeakGroup.msrun_sample` (`ForeignKey`) **to be optional**
- Add: `PeakGroup.sample` (`ForeignKey`) **required**

We know reasonably confidently that a particular peak group is linked to a sample, even if we may not know for sure which `mzXML` it came from.  We can keep `PeakGroup.msrun_sample` around for when we *do* know that association.  And it allows us to be able to retrieve all `mzXML` files associated with a `Sample` from the standpoint of a `PeakGroup`, while maintaining the ability to get back and forth between `Sample` and `PeakGroup`.

This will free us up to be able to load `mzXML` files anytime we want, because we can also reliably associate `mzXML` files with a sample (just not with a `PeakGroup` or `accucor` file.  We no longer will need to create empty placeholder `MSRunSample` records.  We can create a sequence, because we were given that information, but even that can be cumbersome for the user to backtrack and figure out if there were multiple sequences.  Which one was run on which day?  It would be more convenient for the user to work that out when they are ready to submit `mzXML` files.  When they are loading accucor data, working out the sequence(s) just gets in the way.

A problem this *creates* however is, when we don't have mzXML files, but we know the person, day, and protocol, we don't have a way to associate a peak group or accucor file, with that previously loaded accucor data if the person/day/protocol is in fact known (we just don't have the `mzXML`).  To accommodate that, I propose creating:

- `PeakGroup.msrun_sequence` (`ForeignKey`) **optional**

We could make this required, but again, requiring the user to work this out is burdensome and is a hurdle to submission.  We have a researcher and date available via `PeakGroup.sample.researcher` and `PeakGroup.sample.date` which can help to later associate a `PeakGroup`/accucor with a sequence.  That's the work that we're requiring the user to do when submissing an accucor which I think is putting users off.

By not requiring we have or create `MSRunSample` records, it means that we can load `MSRunSequence` and `MSRunSample` records anytime!  We can avoid adding them entirely.  We can add them retroactively.  We can add them when we get them.  We're not dependent on them anymore.

And since samples are unique, from any peakgroup, we can retrieve all mzXMLs with a query of the MSRunSample table using `PeakGroup.sample`.

The only downside is that if we don't have `mzXML` files, we cannot associate a `PeakGroup` with a polarity, researcher, date, protocol, etc.  However, we don't care about polarity.  The researcher, date, and protocol however is metadata we're just taking people's word on anyway, but if we want to preserve that association, I don't think "inventing" a lossy `MSRunSample` record, which is really a placeholder anyway, is helping us at all.  It's just creating problems.

What we reasonably *know* are the relations: `mzXML`:`Sample`, `PeakGroup`:`Sample`.  We don't really (or in every instance) know `mzXML`:`MSRunSequence`, `mzXML`:`accucor`, or `mzXML`:`PeakGroup`, nor do we care that much.

What this means is that the only reason we ever need to create an `MSRunSample` record is when we have an `mzXML` file.  And if we know the specific one that went into a column of accucor data, we can link `PeakGroup`->`MSRunSample` (and the indorect association of `accucor`->`MSRunSample` via `PeakGroup.msrun_sample`.

### Alternatives

None

### Dependencies

None

### Comment

None

-----

### Assumptions

1. List of assumptions that the code will not explicitly address/check
2. E.g. We will assume input is correct (explaining why there is no validation)

### Limitations

1. A list of things this work will specifically not do
2. E.g. This feature will only handle the most frequent use case X

### Affected Components

- change: File path or DB table ...
- add: Environment variable or server setting
- delete: External executable or cron job

### Requirements

- [ ] 1. List of numbered conditions to be met for the feature
- [ ] 2. E.g. Every column/row must display a value, i.e. cannot be empty
- [ ] 3. Numbers for reference & checkboxes for progress tracking

#### Interface Change description

None provided

#### Code Change Description

None provided

#### Tests

- [ ] 1. A description of at least one test for each requirement above.
- [ ] 2. E.g. Test for req 2 that there's an exception when display value is ''
- [ ] 3. Numbers for reference & checkboxes for progress tracking

**Comment by jcmatese** [comment]
>  After I submit this issue, I will proceed directly to implementing what was discussed in recent meetings as high priority: _the ability to load mzXML files unassociated with loaded accucor data_. 

Q: What meeting(s) was this? Who specified the feature, and the priority?  I must have missed that meeting, because I have no recollection of this...

**Comment by lparsons** [comment]
> > After I submit this issue, I will proceed directly to implementing what was discussed in recent meetings as high priority: _the ability to load mzXML files unassociated with loaded accucor data_.
> 
> Q: What meeting(s) was this? Who specified the feature, and the priority? I must have missed that meeting, because I have no recollection of this...

This was feedback directly from Josh. The best written representation at the moment is the description of the milestone for release 3.0: https://github.com/Princeton-LSI-ResearchComputing/tracebase/milestone/24. It was also somewhat alluded to in Josh's comment at the lab meeting about the data for the "historical" studies. He's very keen to get that organized, and I think that is probably going be the goal that drives this feature. For new data, I'm very hopeful that it will be organized in the `msdata` share using unique sample ids that we can reference, and thus, the need to store it with TraceBase will be greatly reduced.

**Comment by hepcat72** [comment]
I believe Lance mentioned it in a recent TraceBase planning meeting?  Maybe he just said it to me over lunch or something, but I'm pretty sure it has been alluded to on more than one meeting occasion, although I don't know if it was ever a focussed topic of discussion?  Regardless, it was made clear to me that that was a high priority goal after the v3 beta release.

**Comment by hepcat72** [comment]
I think the issue is best described in #905.  So the only question is surrounding its priority.

**Comment by hepcat72** [comment]
I updated the inspiration section because I hadn't found #905 when I originally looked for it.

**Comment by hepcat72** [comment]
A convo transferred from slack...

OK.  I'm going to tell a story here with pictures.  Let me get to the end before we discuss.

I like to depict schemas as having a primary trunk/path that has the most important data.  This depicts that path that we have right now, and MSRunSample is an integral part of that backbone, and since we don't always have the data necessary to make it work, there-in lies the source of the problem.  While it accurately conceptualizes reality, it has been a thorn in our side from the beginning, because we could not make all MSRunSample records unique enough for all cases.

![schema_current_backbone](https://github.com/Princeton-LSI-ResearchComputing/tracebase/assets/2300532/83e19071-462d-4e8e-9b4b-1efa16681085)

What I'm proposing is that we essentially have 2 independent paths. We excise the MSRunSample table from the backbone to separate it and make it not necessary to load accucor data. Thus we have our 2 independent paths:

![excise_msrun_sample_720](https://github.com/Princeton-LSI-ResearchComputing/tracebase/assets/2300532/55790946-5801-46ef-bd03-208e0c098547)
![msrunsequence_to_study_720](https://github.com/Princeton-LSI-ResearchComputing/tracebase/assets/2300532/0d29feb1-1d21-468a-bcc1-1a77fb5208fb)

The common/unchanging part is here:

![sample_to_study_720](https://github.com/Princeton-LSI-ResearchComputing/tracebase/assets/2300532/fdccd8d8-8acf-41af-bbb3-9201797710be)

By adding PeakGroup.sample , we now have a branch along the trunk, showing the 2 main and independent paths. This shows that the MSRun and accucor data can be loaded independently (though it doesn't show other links I'm proposing we add, just so I can emphasize the paths.

![branch_720](https://github.com/Princeton-LSI-ResearchComputing/tracebase/assets/2300532/1415cd33-b9c3-4c21-a7ed-d41bda512c74)

But, we want to associate accucor data with the mzXMLs, when we know the association, and we want to retain the association of accucor data with the MS operator, date, protocol, etc when we don't have mzXMLs, so we add PeakGroup.msrun_sequence.  With all the associations, we end up with this:

![final_reframe_720](https://github.com/Princeton-LSI-ResearchComputing/tracebase/assets/2300532/23d4d465-a5ac-4ae3-a86f-21978bcea305)

The result is: Accucor data loading and mzXML data loading are completely independent.  We would make PeakGroup.msrun_sample an optional field. It removes the need for placeholder MSRunSequence records, and all the overhead that that would entail when retro-actively adding mzXML files.  And we'd be able to retrieve all mzXMLs associated with a sample, even via a PeakGroup.

Because the thing is, even if we don't have unique enough data to link PeakGroup to Sample via MSRunSample, we can always accurately link PeakGroup directly to Sample.  And this shows that the only reason we need an MSRunSample record is to associate mzXML files with PeakGroups, accucor files, and Samples.

Anyway, that's the essence of my pitch.  And I think that the modifications are relatively easy and would not involve a database rebuild.  just a migration.

If you guys want to play around with it on QDB, here is the code:
```
# https://app.quickdatabasediagrams.com/#/
#
# Modify this code to update the DB schema diagram.
# To reset the sample schema, replace everything with
# two dots ('..' - without quotes).

MSRunSample
-
id PK int
sample int FK >- Sample.id
# sample int
polarity string
mz_min float
mz_max float
ms_raw_file int
ms_data_file int
msrun_sequence int FK >- MSRunSequence.id

MSRunSequence
-
id PK int
researcher string
date date
instrument string
lc_method int
notes string

Sample
----
id PK int
name string
animal int FK >- Animal.id
tissue int
date date
researcher string

# Table documentation comment 1 (try the PDF/RTF export)
PeakGroup
------------
id PK int
sample int FK >- Sample.id
name string
formula string
compound int
msrun_sequence int FK >- MSRunSequence.id
msrun_sample int FK >- MSRunSample.id
# msrun_sample int
peak_annotation_file int

PeakData
----
id PK int
peak_group int FK >- PeakGroup.id

Animal
----
id int
study int FK >- Study.id

Study
----
id int
```

**Comment by hepcat72** [comment]
Going to go back to the "bucket of files" design, so we will always have a placeholder.
